runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,325 @@
+            log.error("Cannot start new transaction in this thread without ending previous one");
+            throw new TransactionAlreadyStartedException(TransactionalContext.getRootContext().toString());
+        }
+    }
+
+    public  <K extends Message, V extends Message, M extends Message>
+    Table<K, V, M> getTable(@Nonnull final String tableName) {
+        return this.tableRegistry.getTable(this.namespace, tableName);
+    }
+
+    /**
+     *************************** WRITE APIs *****************************************
+     */
+
+    /**
+     * put the value on the specified key create record if it does not exist.
+     *
+     * @param table Table object to perform the create/update on.
+     * @param key   Key of the record.
+     * @param value Value or payload of the record.
+     * @param metadata Metadata associated with the record.
+     * @param <K>   Type of Key.
+     * @param <V>   Type of Value.
+     * @param <M>   Type of Metadata.
+     * @return TxnContext instance.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message>
+    TxnContext put(@Nonnull Table<K, V, M> table,
+                   @Nonnull final K key,
+                   @Nonnull final V value,
+                   @Nullable final M metadata) {
+        operations.add(() -> {
+            table.put(key, value, metadata);
+        });
+        return this;
+    }
+
+    /**
+     * Merges the delta value with the old value by applying a caller specified BiFunction and writes
+     * the final value.
+     *
+     * @param table Table object to perform the merge operation on.
+     * @param key            Key
+     * @param mergeOperator  Function to apply to get the new value
+     * @param recordDelta    Argument to pass to the mutation function
+     * @param <K>            Type of Key.
+     * @param <V>            Type of Value.
+     * @param <M>            Type of Metadata.
+     * @return TxnContext instance
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message>
+    TxnContext merge(@Nonnull Table<K, V, M> table,
+                     @Nonnull final K key,
+                     @Nonnull BiFunction<CorfuRecord<V, M>, CorfuRecord<V,M>, CorfuRecord<V,M>> mergeOperator,
+                     @Nonnull final CorfuRecord<V,M> recordDelta) {
+        operations.add(() -> {
+            CorfuRecord<V,M> oldRecord = table.get(key);
+            CorfuRecord<V, M> mergedRecord;
+            try {
+                mergedRecord = mergeOperator.apply(oldRecord, recordDelta);
+            } catch (Throwable throwable) {
+                throw new TransactionAbortedException(
+                        new TxResolutionInfo(table.getStreamUUID(), this.isolationLevel.getTimestamp()),
+                        AbortCause.USER,
+                        throwable,
+                        TransactionalContext.getCurrentContext());
+            }
```

Comment:
I think this exception can be handled by user, instead of we catching. Think about a normal put() operation, which could also throw an exception. User should call our API in a try catch finally block.